docs/core/knowledge-bases.rst

```diff
@@ -0,0 +1,566 @@
+          "3": lambda l: l[2],
+          "4": lambda l: l[3],
+          "5": lambda l: l[4],
+          "6": lambda l: l[5],
+          "7": lambda l: l[6],
+          "8": lambda l: l[7],
+          "9": lambda l: l[8],
+          "10": lambda l: l[9],
+          "ANY": lambda l: random.choice(list),
+          "LAST": lambda l: l[-1],
+      }
+
+  You can overwrite it by calling the function ``set_ordinal_mention_mapping``.
+  If you want to learn more about the usage of the mapping, go to section :ref:`resolve_mentions`.
+
+
+
+Customize your `ActionQueryKnowledgeBase`
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+You can overwrite the following two functions of `ActionQueryKnowledgeBase`:
+
+- ``utter_objects()``
+- ``utter_attribute_value()``
+
+``utter_objects()`` is used when the user requested the bot to list some objects.
+Once the bot retrieved some objects from the knowledge base, it will response to the user, for example, with
+
+    `Found the following objects of type 'restaurant':`
+    `1: I due forni`
+    `2: PastaBar`
+    `3: Berlin Burrito Company`
+
+Or if no entities could be found
+
+    `I could not find any objects of type 'restaurant'.`
+
+If you want to change the utterance of the bot, you can overwrite the method ``utter_objects()`` in your action.
+
+The function ``utter_attribute_value()`` determines what the bot utters when the user is asking for a detail of
+an object.
+If the attribute of interest was found in the knowledge base, the bot will response with the following utterance:
+
+    `'Berlin Burrito Company' has the value 'Mexican' for attribute 'cuisine'.`
+
+If no value for the requested attribute was found, the bot will response with
+
+    `Did not found a valid value for attribute 'cuisine' for object 'Berlin Burrito Company'.`
+
+If you want to change the utterance of the bot, you can overwrite the method ``utter_attribute_value()``.
+
+.. note::
+   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
+
+
+.. _custom_knowledge_base:
+
+Creating your own Knowledge Base
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you have more data or if you want to use a more complex data structure that, for example, involves relations between
+different objects, you can create your own knowledge base implementation.
+Just inherit ``KnowledgeBase`` and implement the methods ``get_objects()``, ``get_object()``, and
+``get_attributes_of_object()``.
+You can also customize your knowledge base further, for example, by adapting the methods mentioned in the section
+:ref:`customize_in_memory_knowledge_base`.
+
+.. note::
+   We wrote a `blog post <https://blog.rasa.com/set-up-a-knowledge-base-to-encode-domain-knowledge-for-rasa/>`_
+   that explains how you can set up your own knowledge base.
+
+
+Creating your own Knowledge Base Actions
```

Comment:
Should probably go above creating your own knowledge base -- feel like the level of customisation would make most sense as

1. Customize actionqueryknowledgebase 
2. Create your own knowledge base action 
3. Customize the in memory knowledge base 
4. Create your own knowledge base